vim: Fix `ap` text object selection when there is line wrapping

vieiralucas opened the pull request:
In Vim mode, `ap` text object (used in `vap`, `dap`, `cap`) was selecting multiple paragraphs when soft wrap was enabled. The bug was caused by using DisplayRow coordinates for arithmetic instead of buffer row coordinates in the paragraph boundary calculation.

Fix by converting to buffer coordinates before arithmetic, then back to display coordinates for the final result.

Closes #35085

vieiralucas committed:
fix(vim): `ap` text object selection when there is line wrapping

In Vim mode, `ap` text object (used in `vap`, `dap`, `cap`) was selecting
multiple paragraphs when soft wrap was enabled. The bug was caused by using
DisplayRow coordinates for arithmetic instead of buffer row coordinates in
the paragraph boundary calculation.

Fix by converting to buffer coordinates before arithmetic, then back to
display coordinates for the final result.

vieiralucas commented:
Do you think some tests should be added for this? I want to do that but still learning the codebase so will not be easy 🙈.

ConradIrwin commented:
@vieiralucas Would love a test for this! You should be able to copy-paste one of the existing vim tests like `test_wrapped_delete_end_document` that uses `set_shared_wrap`. To generate the JSON file you need to run `cargo test -p vim --features neovim TEST_NAME`

vieiralucas committed:
fix(vim): preserve cursor column when deleting paragraphs with `dap`

When deleting a paragraph with the "around" modifier (dap command),
the cursor now attempts to maintain its original column position
after deletion, matching vim's behavior. This only applies when
the current line contains non-whitespace characters.

Previously, the cursor would always move to the beginning of the line
or follow different positioning logic that didn't match vim's expectations.

vieiralucas committed:
fix(vim): paragraph object cursor positioning in visual mode

This commit addresses cursor positioning issues in Zed's Vim emulation for
paragraph objects (v i p and v a p) to match Neovim's behavior.

Key Issues Fixed:

Cursor Positioning Logic:
  - Single blank line + inner paragraph (v i p): cursor at START of selection
  - All other cases (multi-line paragraphs, around paragraphs): cursor at END

Empty Range Handling:
  - Fixed issue where empty ranges (single blank line paragraphs) weren't applied
  - Moved cursor positioning logic outside the !range.is_empty() condition
  - Ensured empty ranges result in proper visual line selections

Visual Line Mode Selection:
  - For empty ranges, expand selection to span full line
  - For non-empty ranges, include newline by extending to start of next line
  - Use buffer coordinates consistently for paragraph logic

Testing Framework Improvements:

Enhanced Neovim Integration:
  - Fixed visual mode selection capture using '< and '> marks
  - Added proper handling for visual mode state transitions during test recording
  - Improved selection bounds detection for more accurate test comparisons

Test Coverage:
  - Added comprehensive test cases for single/multi-line blank paragraphs
  - Added soft wrap paragraph tests to ensure display coordinate handling
  - All paragraph-related tests now pass with correct cursor positioning

This ensures Zed's paragraph object behavior matches Neovim exactly,
particularly for edge cases involving blank lines and cursor positioning
in visual line mode.

vieiralucas commented:
Hey @ConradIrwin, so this turned out to be harder than anticipated.

While adding tests for the original paragraph text object fix, I discovered several additional issues that needed to be addressed:

#### Test Framework Issues
- The Neovim integration was incorrectly capturing visual selections by reconstructing them from cursor and visual mark positions
- Fixed by using `'<` and `'>` marks to get the actual selection bounds from Neovim
- This was causing visual mode tests to pass when they should have been failing

#### Paragraph Object Visual Mode Issues  
- Cursor positioning didn't match Neovim behavior:
  - Single blank line inner paragraphs (`v i p`) should have cursor at START
  - All other cases should have cursor at END
- Empty ranges (single blank line paragraphs) weren't being applied at all due to logic being inside `!range.is_empty()` check
- Visual line mode wasn't properly selecting full lines for edge cases

#### Changes Made
- Updated `visual.rs` to handle paragraph objects specially with correct cursor positioning logic
- Fixed empty range handling to ensure single blank lines get selected properly  
- Enhanced test framework in `neovim_connection.rs` to capture visual selections accurately
- Added comprehensive test coverage for various paragraph object scenarios

These fixes ensure paragraph objects now match Neovim's behavior exactly and that our test framework can properly validate visual mode behavior.

See the individual commit messages for more detailed explanations of each fix.

vieiralucas commented:
Hey @ConradIrwin, so this turned out to be harder than anticipated.

While adding tests for the original paragraph text object fix, I discovered several additional issues that needed to be addressed:

#### Test Framework Issues
- The Neovim integration was incorrectly capturing visual selections by reconstructing them from cursor and visual mark positions
- Fixed by using `'<` and `'>` marks to get the actual selection bounds from Neovim
- This was causing visual mode tests to pass when they should have been failing


#### Delete Paragraph Cursor Positioning
- After deleting paragraphs with `dap`, the cursor wasn't preserving its original column position
- Fixed to match Vim's behavior: attempt to maintain column position when the current line has non-whitespace characters

#### Paragraph Object Visual Mode Issues  
- Cursor positioning didn't match Neovim behavior:
  - Single blank line inner paragraphs (`v i p`) should have cursor at START
  - All other cases should have cursor at END
- Empty ranges (single blank line paragraphs) weren't being applied at all due to logic being inside `!range.is_empty()` check
- Visual line mode wasn't properly selecting full lines for edge cases

#### Changes Made
- Added cursor column preservation logic in `delete.rs` for paragraph deletion
- Updated `visual.rs` to handle paragraph objects specially with correct cursor positioning logic
- Fixed empty range handling to ensure single blank lines get selected properly  
- Fixed test framework in `neovim_connection.rs` to capture visual selections accurately
- Added test coverage for various paragraph object scenarios including soft wrap cases

These fixes ensure paragraph objects now match Neovim's behavior exactly and that our test framework can properly validate visual mode behavior.

See the individual commit messages for more detailed explanations of each fix.

vieiralucas commented:
It's all good, don't get me wrong. It was an interesting and fun journey 😸

About the test framework thing, maybe my fix is not the best, but the thing is that it was causing false positives in our tests:

1. Zed would produce the correct visual selection 
2. Neovim would also produce the correct visual selection
3. But our test framework would capture the wrong bounds from Neovim
4. The test would compare Zed's correct selection against Neovim's incorrectly-captured selection
5. If they happened to match the wrong bounds, tests would pass even when Zed was incorrect

We even had a test (`test_v2ap`) that was hardcoded instead of using `assert_matches` - likely because the framework couldn't capture the right expected behavior! 🤣 
https://github.com/zed-industries/zed/pull/35485/files#diff-43f534758348a459822cb280c370a35979304dc76bcb078ce6396f08f6dadf1bL1802-L1808

I understand your point about VISUAL LINE mode - you're right that the whole lines look selected regardless of internal bounds. The issue I was trying to solve was that the old framework was using cursor position + visual mark instead of the actual selection bounds that Neovim computed, which gave wrong coordinates even for VISUAL LINE selections.

But you designed the original framework, so you definitely know the trade-offs better than I do. I'm happy to revert the test framework changes if you think the previous approach was better, or adjust it however you'd prefer.

yutayamamoto commented:
Perhaps the `yap` case is missing? (I checked out this PR to see if it's related to issue #36171 and manually tested it myself, and it seems that the bug reported in #28804 for `yap` is still there, while `vap`, `dap`, and `cap` are correctly fixed.)

ConradIrwin committed:
Revert change to vim test infra

ConradIrwin committed:
Revert change to visual paragraph selection

ConradIrwin committed:
Better fix

ConradIrwin committed:
Fix clipping in delete mode

ConradIrwin committed:
Test

ConradIrwin commented:
@vieiralucas `assert_matches!` does `assert_eq!` under the hood (I like having the assertion in the source-code for debugging, even though it's superfluous).

I think your change to the test framework is not right, because it loses the order of the selection (`'<` and `'>` are always the start and end of the selection, but in real life, the selection can be made from start-to-end or end-to-start (and `o` toggles between the two).

The other problem is that in both vim and Zed, the underlying selection is always character wise, and the line-mode stuff is overlaid on. If you do `vip` in Vim, you see the whole line selected, but if you then press `v` you'll see that the underlying selection is only the first character, and it's being extended by line mode.

I think your changes to preserve columns in deletion are right (but they should apply for any line-wise deletion, including when the line is blank). I also think the changes to soft-wrap handling in the range function of paragraph are good. But the change to the selection range for visual mode in paragraphs isn't quite right.

I've reverted those things, and added some more code to handle the edge case you found in test_visual_paragraph_object_with_soft_wrap, where the tail of the selection is preserved when selecting a line-wise object if the selection started on the first line of the object.

Are there other things you noticed that we also need to fix? Otherwise I think this is good to go (assuming the tests pass).

vieiralucas commented on crates/vim/src/object.rs:
```diff
@@ -1903,6 +1904,95 @@ mod test {
+    async fn test_delete_paragraph_whitespace(cx: &mut gpui::TestAppContext) {
+        let mut cx = NeovimBackedTestContext::new(cx).await;
+
+        // Test paragraphs that will wrap when soft wrap is enabled
+        cx.set_shared_state(indoc! {"
+            a
+                   ˇ•
+            aaaaaaaaaaaaa
+        "})
+            .await;
+
+        cx.simulate_shared_keystrokes("d i p").await;
+        // Test paragraphs that will wrap when soft wrap is enabled
```
@ConradIrwin 
Are these comments wrong? Or did you forgot to call `set_shared_wrap`?

vieiralucas commented:
Hi @ConradIrwin,

> ... the selection can be made from start-to-end or end-to-start (and o toggles between the two). 

> ... but if you then press v you'll see that the underlying selection is only the first character, and it's being extended by line mode.

Did not know about those, that makes a lot of sense, thank you for taking the time to explain.

<br />
> Are there other things you noticed that we also need to fix?

No
<br />
I left one comment on your changes, the `test_delete_paragraph_whitespace` has a comment talking about soft wrap but the test does not seems to actually be enabling soft wrap?
<br />
Sorry if I ended up creating more work for you.

ConradIrwin committed:
Remove comments

ConradIrwin commented:
No need to be sorry, thank you for showing up and fixing things! Makes it easier for me to justify spending time on the edge cases :D.
